framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4631,6 +4622,450 @@ private void makeGroundTargetType(
+    TypeMirror capturedTypeMirror = types.capture(typeToCapture.getUnderlyingType());
+    return applyCaptureConversion(typeToCapture, capturedTypeMirror);
+  }
+
+  /**
+   * Apply capture conversion to {@code type}.
+   *
+   * <p>Capture conversion is the process of converting wildcards in a parameterized type to fresh
+   * type variables. See <a
+   * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-5.html#jls-5.1.10">JLS 5.1.10</a>
+   * for details.
+   *
+   * <p>If {@code type} is not a declared type or if it does not have any wildcard type arguments,
+   * this method returns {@code type}.
+   *
+   * @param type type to capture
+   * @param typeMirror the result of applying capture conversion to the underlying type of {@code
+   *     type}; it is used as the underlying type of the returned type
+   * @return the result of applying capture conversion to {@code type}
+   */
+  public AnnotatedTypeMirror applyCaptureConversion(
+      AnnotatedTypeMirror type, TypeMirror typeMirror) {
+
+    // If capture conversion the type contains uninferred type arguments, don't capture, but
+    // mark all wildcards in type as "uninferred" before it is returned.
+    if (type.containsUninferredTypeArguments()
+        && typeMirror.getKind() == TypeKind.DECLARED
+        && type.getKind() == TypeKind.DECLARED) {
+      AnnotatedDeclaredType uncapturedType = (AnnotatedDeclaredType) type;
+      DeclaredType capturedTypeMirror = (DeclaredType) typeMirror;
+      for (int i = 0; i < capturedTypeMirror.getTypeArguments().size(); i++) {
+        AnnotatedTypeMirror uncapturedTypeArg = uncapturedType.getTypeArguments().get(i);
+        TypeMirror capturedTypeArgTM = capturedTypeMirror.getTypeArguments().get(i);
+        if (uncapturedTypeArg.getKind() == TypeKind.WILDCARD
+            && (TypesUtils.isCapturedTypeVariable(capturedTypeArgTM)
+                || capturedTypeArgTM.getKind() != TypeKind.WILDCARD)) {
+          ((AnnotatedWildcardType) uncapturedTypeArg).setUninferredTypeArgument();
+        }
+      }
+      return type;
+    }
+
+    if (!shouldCapture(type, typeMirror)) {
+      return type;
+    }
+
+    AnnotatedDeclaredType uncapturedType = (AnnotatedDeclaredType) type;
+    DeclaredType capturedTypeMirror = (DeclaredType) typeMirror;
+    // `capturedType` is the return value of this method.
+    AnnotatedDeclaredType capturedType =
+        (AnnotatedDeclaredType) AnnotatedTypeMirror.createType(capturedTypeMirror, this, false);
+
+    nonWildcardTypeArgCopier.copy(uncapturedType, capturedType);
+
+    AnnotatedDeclaredType typeDeclaration =
+        (AnnotatedDeclaredType) getAnnotatedType(uncapturedType.getUnderlyingType().asElement());
+
+    // A mapping from type variable to its type argument in the captured type.
+    Map<TypeVariable, AnnotatedTypeMirror> typeVarToAnnotatedTypeArg = new HashMap<>();
+    // A mapping from a captured type variable to the annotated captured type variable.
```

Comment:
Is the key from the underlying type or from the captured type?

Reply:
I don't understand this question.  Which underlying type do you mean?